server/models/procedures.sql

```diff
@@ -18,6 +18,235 @@ BEGIN
 END
 $$
 
+CREATE PROCEDURE StageInvoice(
+  IN is_distributable TINYINT(1),
+  IN date DATETIME,
+  IN cost DECIMAL(19, 4) UNSIGNED,
+  IN description TEXT,
+  IN service_id SMALLINT(5) UNSIGNED,
+  IN debtor_uuid BINARY(16),
+  IN project_id SMALLINT(5),
+  IN user_id SMALLINT(5),
+  IN uuid BINARY(16)
+)
+BEGIN 
+  DECLARE `no_invoice_stage` TINYINT(1) DEFAULT 0;
+  DECLARE CONTINUE HANDLER FOR SQLSTATE '42S02' SET `no_invoice_stage` = 1;
+  SELECT NULL FROM `stage_invoice` LIMIT 0;
+
+  IF (`no_invoice_stage` = 1) THEN 
+    create temporary table stage_invoice 
+    (select project_id, uuid, cost, debtor_uuid, service_id, user_id, date, description, is_distributable);
+  ELSE 
+    insert into stage_invoice
+    (select project_id, uuid, cost, debtor_uuid, service_id, user_id, date, description, is_distributable);
+  END IF;
+END $$
+
+CREATE PROCEDURE StageInvoiceItem(
```

Comment:
This is fascinating, by the way.  Really impressive that this works!  It looks pretty clean, even if it is an order of complexity above simple transactions.